### ISSUE TITLE
JMAP protocol support

### ISSUE THREAD

**Issue by okias** [k9mail/k-9#3272]
I'd like to create this issue for tracking of JMAP protocol support [1] progress/interest. I guess at some point JMAP protocol will be implemented by K9 Mail.

This time, JMAP is already implemented in few Mail servers, so wide adoption will probably follow soon.

[1] http://jmap.io/

**Comment by AvoidMastodonBlue** [comment]
> IMAP is either the woefully inadequate original RFC3501, or it’s a messy set of incomplete implementations of some of the extensions


Why bother correctly implementing? Create new standard and magical compliance?

**Comment by okias** [comment]
As I understood, since numerous extensions on a top of really old protocol is really hard to make it work as it should. Otherwise I cannot explain, why there is so small number of (well)working emails clients. And even clients as Thunderbird slowly going into maintenance mode, since no-one wants develop something more sophisticated a top of IMAP protocol.

**Comment by tya99** [comment]
Recently this was published https://fastmail.blog/2018/12/27/jmap-is-on-the-home-straight/

I think JMAP is an eventual evolution.  It has a lot to offer particularly on mobile devices which IMAP was not designed for. It would be a suitable investment in time for K-9 to be the first JMAP client. I do see this being eventually baked into Android however as a complete replacement for IMAP.

**Comment by marclaporte** [comment]
This request (which I strongly support) is now listed at: https://jmap.io/software.html

**Comment by tya99** [comment]
> It was considered:
> https://github.com/k9mail/k-9/wiki/Google-Summer-of-Code-2017

Ah.

If they ever have another GSOC I think sending email with `format=flowed` would be good. Sad to see it's unwanted #2302 I really like how Thunderbird does it. Note `quoted-printable` is still used for pgp signed email.

**Comment by iNPUTmice** [comment]
Just as a quick FYI there is a Java / Android library available here: https://github.com/iNPUTmice/jmap

The *jmap-mua* part of the library makes a lot of assumptions about the mail client, its work flow and how it stores it's data and might not be a good fit for k9. However the lower level *jmap-client* module would still take a lot of work of the shoulders of anyone wanting to implement JMAP in k9.

**Comment by yoshimo** [comment]
How well is this supported by providers these days?

**Comment by cketti** [comment]
This is an issue tracker to support the development of the app. Please use the [mailing list](http://groups.google.com/group/k-9-mail) for general discussions.

**Comment by bortzmeyer** [comment]
The standard has just been published: RFC 8621 https://www.rfc-editor.org/info/rfc8621

**Comment by cketti** [comment]
Short update: The project funding was approved and I'll be working on this for the next 6 months.

**Comment by Pete879739** [comment]
K-9 Mail has not been reliably syncing new emails with FastMail (via IMAP) ever since FastMail turned on the JMAP protocol.  Sometimes I have to manually refresh K-9 Mail.  Maybe it's just a coincidence though.

**Comment by marclaporte** [comment]
@cketti The great news was published here: https://jmap.io/news.html#k-9-mail-is-working-on-adding-jmap-support

If you have any updates, we are very interested :-)

Thanks!

**Comment by cketti** [comment]
The 6 months are up. The repository now contains a module called [k9mail-jmap](https://github.com/k9mail/k-9/tree/ecec1e1db5d0261646d3d654eb6b82760a0a118d/app/k9mail-jmap) that is a version of the K-9 Mail app with rudimentary support for JMAP. The usability is not great and I don't plan on making APKs of it available. People interested in the current state will have to build the app themselves.

I created a [JMAP milestone](https://github.com/k9mail/k-9/milestone/26) in the issue tracker with things that have to be implemented before I feel comfortable to include the JMAP code in the main app.

I plan to continue working on JMAP support, but I don't consider it a blocker for the next stable version of K-9 Mail. So it might be a while before I'll get back to it.

**Comment by okias** [comment]
Almost one year passed. Except funded work, I assume there is no motivation to implement it into usable form?

Leaving out JMAP leave us with stagnation or implementing Microsoft Exchange protocol?

**Comment by SRoyPC** [comment]
What is the current status?

**Comment by bam80** [comment]
One year more.

**Comment by Mhsbrian** [comment]
Has this now became a beyond request?

**Comment by doctorlard** [comment]
This is something I would like to contribute to; I'm considering moving (some of) my mail to Fastmail, so I can test it. This should probably depend on the updated UnifiedPush issue #5165 (which is currently understandably locked as off-topic)

**Comment by cketti** [comment]
There's a milestone for adding JMAP support: https://github.com/thundernest/k-9/milestone/26

It's currently not a high priority and there's a significant amount of work left to get this into a state that is good enough to be included in the app. 
If you want to contribute to this, please try to keep the pull requests at a reasonable size and don't try to make too many changes at once.

UnifiedPush is just one possible push mechanism that can be used with JMAP. It's an optional feature, not a requirement.

**Comment by quickfixdemon** [comment]
What's the status on JMAP support?

**Comment by kewisch** [comment]
JMAP support is not on our immediate roadmap, therefore this unfortunately isn't something we'll be getting to soon. Given the scope of this project it might be worth creating an idea on https://mzla.link/tb-android-connect that others can upvote. I'd love to see broader adoption of JMAP from email providers to support the urgency as well.

**Comment by r0bbie** [comment]
Like others am hopeful to see JMAP support become a priority in future - we really need a solid mobile client.

There's an existing idea here - https://connect.mozilla.org/t5/ideas/jmap-support-or-similar-for-thunderbird/idi-p/28974 for Thunderbird, I guess a separate one should be setup for support in K9?

**Comment by kewisch** [comment]
Yes, a separate idea for Android would be great, you can link to the other idea (and this issue) and the connect folks will do the right things. Also note that it doesn't need to be on the Thunderbird for Android roadmap for others to work on. If someone want to take a few small steps towards JMAP they are certainly welcome to!

**Comment by yan12125** [comment]
> The 6 months are up. The repository now contains a module called [k9mail-jmap](https://github.com/k9mail/k-9/tree/ecec1e1db5d0261646d3d654eb6b82760a0a118d/app/k9mail-jmap) that is a version of the K-9 Mail app with rudimentary support for JMAP.

The k9mail-jmap module is removed in https://github.com/thunderbird/thunderbird-android/pull/5872, and there is no screen to configure JMAP server settings per https://github.com/thunderbird/thunderbird-android/issues/4566. Does that mean there is no way to test JMAP functionality with the current codebase, considering the aporoach of building the app from sources?

**Comment by Drab3229** [comment]
any action on this?
